RNWCPP/Chakra/ChakraJsiRuntime_core.cpp

```diff
@@ -0,0 +1,318 @@
+}
+
+void ChakraJsiRuntime::setupNativePromiseContinuation() noexcept{
+  if (args_.promiseContinuation) {
+    JsSetPromiseContinuationCallback(PromiseContinuationCallback, this);
+  }
+}
+
+// This is very wierd. This should match with the definition of VS_VERSIONINFO as defined in https://docs.microsoft.com/en-us/windows/desktop/menurc/vs-versioninfo
+// I can't find a way to include the actual definition of VS_VERSIONINFO
+// TODO :: Re-evaluate this strategy.
+struct FileVersionInfoResource {
+  uint16_t len;
+  uint16_t valLen;
+  uint16_t type;
+  wchar_t key[_countof(L"VS_VERSION_INFO")];
+  uint16_t padding1;
+  VS_FIXEDFILEINFO fixedFileInfo;
+  uint32_t padding2;
+};
+
+// TODO :: This code is mostly copied from the old ChakraExecutor flow, and not verified for reliability yet.
+// TODO :: Re-evaluate this strategy of finding the dll version for versioning the runtime.
+/*static*/ void ChakraJsiRuntime::initRuntimeVersion()  noexcept {
+  // This code is win32 only at the moment. We will need to change this
+  // line if we want to support UWP.
+  constexpr wchar_t chakraDllName[] = L"ChakraCore.dll";
+
+  auto freeLibraryWrapper = [](void* p) { FreeLibrary((HMODULE)p); };
+  HMODULE moduleHandle;
+  if (!GetModuleHandleExW(0, chakraDllName, &moduleHandle))
+    return;
+
+  std::unique_ptr<void, decltype(freeLibraryWrapper)> moduleHandleWrapper{ moduleHandle, std::move(freeLibraryWrapper) };
+
+  HRSRC versionResourceHandle = FindResourceW(moduleHandle, MAKEINTRESOURCE(VS_VERSION_INFO), RT_VERSION);
+
+  if (!versionResourceHandle || SizeofResource(static_cast<HMODULE>(moduleHandleWrapper.get()), versionResourceHandle) < sizeof(FileVersionInfoResource))
+    return;
+
+  HGLOBAL versionResourcePtrHandle = LoadResource(moduleHandle, versionResourceHandle);
+  if (!versionResourcePtrHandle)
+    return;
+
+  FileVersionInfoResource* chakraVersionInfo = static_cast<FileVersionInfoResource*>(LockResource(versionResourcePtrHandle));
+  if (!chakraVersionInfo)
+    return;
+
+  s_runtimeVersion = chakraVersionInfo->fixedFileInfo.dwFileVersionMS;
+  s_runtimeVersion <<= 32;
+  s_runtimeVersion |= chakraVersionInfo->fixedFileInfo.dwFileVersionLS;
+}
+
+#ifdef CHAKRART_HAS_DEBUGGER
+class ChakraJsiRuntimeWithDebugger : public ChakraJsiRuntime {
+public:
+  ChakraJsiRuntimeWithDebugger(ChakraJsiRuntimeArgs&& args);
+  ~ChakraJsiRuntimeWithDebugger();
+
+private:
+  JsErrorCode enableDebugging(JsRuntimeHandle runtime, std::string const& runtimeName, bool breakOnNextLine, uint16_t port, std::unique_ptr<DebugProtocolHandler>& debugProtocolHandler, std::unique_ptr<DebugService>& debugService);
+  void ProcessDebuggerCommandQueue();
+
+  static void CHAKRA_CALLBACK ProcessDebuggerCommandQueueCallback(void* callbackState);
+
+  constexpr static char DebuggerDefaultRuntimeName[] = "runtime1";
+  constexpr static int DebuggerDefaultPort = 9229;
+
+  std::string debugRuntimeName_;

```

Comment:
I have not seen this convention used of the trailing underscore. I don't see the need for it.